In [1]:
import numpy as np
import pandas as pd

In [2]:
from sklearn import preprocessing
from sklearn.feature_selection import RFE
from sklearn.ensemble import RandomForestRegressor
from sklearn.preprocessing import LabelEncoder

In [6]:
data_new = pd.read_csv("C:Data.csv")

In [7]:
data_new.head()

symboling fueltype      Company aspiration  doornumber      carbody  \
0          3      gas  alfa-romero        std           2  convertible   
1          3      gas  alfa-romero        std           2  convertible   
2          1      gas  alfa-romero        std           2    hatchback   
3          2      gas         audi        std           4        sedan   
4          2      gas         audi        std           4        sedan   

  drivewheel enginelocation  wheelbase  carlength  ...  cylindernumber  \
0        rwd          front       88.6      168.8  ...               4   
1        rwd          front       88.6      168.8  ...               4   
2        rwd          front       94.5      171.2  ...               6   
3        fwd          front       99.8      176.6  ...               4   
4        4wd          front       99.4      176.6  ...               5   

   enginesize  fuelsystem boreratio  stroke  compressionratio horsepower  \
0         130        mpfi      3.47    2.68               9.0        111   
1         130        mpfi      3.47    2.68               9.0        111   
2         152        mpfi      2.68    3.47               9.0        154   
3         109        mpfi      3.19    3.40              10.0        102   
4         136        mpfi      3.19    3.40               8.0        115   

   peakrpm    price  fueleconomy  
0     5000  13495.0        23.70  
1     5000  16500.0        23.70  
2     5000  16500.0        22.15  
3     5500  13950.0        26.70  
4     5500  17450.0        19.80  

[5 rows x 24 columns]

In [21]:
data_new = pd.read_csv("C:Data.csv")

X = data_new.apply(lambda col: preprocessing.LabelEncoder().fit_transform(col))
X=X.drop(['Company','price'],axis=1)
y=data_new['price']

# Create the RFE object and rank each pixel
from sklearn.ensemble import RandomForestRegressor
from sklearn.feature_selection import RFE
clf_rf_3 = RandomForestRegressor()      
rfe = RFE(estimator=clf_rf_3, n_features_to_select=5, step=1)
rfe = rfe.fit(X, y)
print('Chosen best 5 feature by rfe:',X.columns[rfe.support_])

features=list(X.columns[rfe.support_])

# Lets create a new Dataframe with Our best Independent Features and Target Varibale
df = data_new[features]
df['price'] = data_new['price']

df = pd.get_dummies(df, drop_first = True)


# Split our data into X and Y

X = df.drop('price',axis=1)
y = df.price

### Train-Test Split

from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test = train_test_split(X, y,
                                                 test_size=0.33, 
                                                 random_state = 45)

Chosen best 5 feature by rfe: Index(['carwidth', 'curbweight', 'enginesize', 'horsepower', 'fueleconomy'], dtype='object')


C:\Users\abhis\AppData\Local\Temp\ipykernel_14048\3109345029.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price'] = data_new['price']


In [22]:
from sklearn.linear_model import LinearRegression

lreg = LinearRegression()
lreg.fit(x_train,y_train)
y_train_pred = lreg.predict(x_train)
y_test_pred = lreg.predict(x_test)
lreg.score(x_test,y_test)

0.7770484304482372

In [23]:
from sklearn.tree import DecisionTreeRegressor

dt_regressor = DecisionTreeRegressor(random_state=0)
dt_regressor.fit(x_train,y_train)
y_train_pred = dt_regressor.predict(x_train)
y_test_pred = dt_regressor.predict(x_test)
dt_regressor.score(x_test,y_test)

0.8419795185060905

In [24]:
from sklearn.ensemble import RandomForestRegressor

Rf = RandomForestRegressor(n_estimators = 15,
                              criterion = 'mse',
                              random_state = 20,
                              n_jobs = -1)
Rf.fit(x_train,y_train)
Rf_train_pred = Rf.predict(x_train)
Rf_test_pred = Rf.predict(x_test)


Rf.score(x_test, y_test)

D:\Installed Software\anaconda3\lib\site-packages\sklearn\ensemble\_forest.py:400: FutureWarning: Criterion 'mse' was deprecated in v1.0 and will be removed in version 1.2. Use `criterion='squared_error'` which is equivalent.
  warn(


0.9123310697942647

Top 5 -  lr = 0.77, DT = 0.84, RF = 0.91
Top 10 - lr = 0.80, DT = 0.81, RF = 0.91
Top 16 - lr = 0.86, DT = 0.82, RF = 0.89